funfact/quantum/_qgates.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from abc import ABC, abstractmethod
+from numbers import Integral
+from typing import List
+from funfact import active_backend as ab
+import numpy as np
+from funfact.parametrized import Generator
+from funfact.lang._tsrex import TsrEx, tensor
+from funfact.lang._ast import Primitives as P
+from funfact.lang._terminal import ParametrizedAbstractTensor
+from funfact.lang._special import eye
+from ._special import proj0, proj1
+from funfact.conditions import Unitary
+# TODO: how to handle dtype? Should always be complex!
+
+
+class Gate(ABC):
+    '''A common base class for all gate objects.'''
+    def __init__(self, qubits: List[Integral]):
+        '''A gate stores the minimum and maximum qubits it acts on.'''
+        for i, q in enumerate(qubits):
+            if not (isinstance(q, Integral) and q >= 0):
+                raise RuntimeError(
+                    f"Qubit indices must be non-negative integers, got {q} for"
+                    f" entry {i}."
+                )
+        qubits.sort()
+        self._qubits = qubits
+
+    @abstractmethod
+    def to_tensor(self):
+        '''A gate can be converted to a tensor, either parametrized or not.'''
+        pass
+
+    @abstractmethod
+    def label(self):
+        '''A gate has a label.'''
+        pass
+
+    @property
+    def qubits(self):
+        return self._qubits
+
+
+class OneQubitGate(Gate):
+    '''Abstract base class for gates acting on exactly one qubit.'''
+    def __init__(self, qubit: Integral):
+        if not (isinstance(qubit, Integral)):
+            raise RuntimeError(
+                f"Qubit index of one qubit gate must be integer, got {qubit}."
+            )
+        super().__init__(list((qubit,)))
+
+
+class OneQubitUnitary(OneQubitGate):
+    '''A dense one qubit unitary gate.'''
+    def __init__(self, qubit: int, initializer=None):
+        self._initializer = initializer
+        super().__init__(qubit)
+
+    def to_tensor(self):
+        return tensor(
+            self.label, 2, 2, initializer=self._initializer, prefer=Unitary()
+        )
+
+    @property
+    def label(self):
+        return 'Q_1'
```

Comment:
Perhaps add an optional argument in the constructor to let users customize the label?